src/main/java/com/researchspace/webapp/integrations/dmponline/DMPOnlineController.java

```diff
@@ -129,47 +288,101 @@ public AjaxReturnObject<JsonNode> importDmp(
-        || StringUtils.isEmpty(dmpOnlineCredentialsAuthorizationCode)) {
-      return null;
+  private String getExistingAccessToken(Model model, Principal principal)
+      throws URISyntaxException, MalformedURLException {
+    Optional<UserConnection> optUserConnection = getExistingUserConnection(principal.getName());
+    if (optUserConnection.isEmpty()) {
+      throw new HttpClientErrorException(HttpStatus.NOT_FOUND, "User connection not found");
     }
-    HttpHeaders headers = new HttpHeaders();
-    headers.add("Accept", "application/x-www-form-urlencoded;charset=UTF-8");
-    headers.add("Content-Type", "application/json");
-    HttpEntity<?> entity =
-        new HttpEntity<>(
-            "{\"grant_type\":\"authorization_code\","
-                + "\"email\":\""
-                + dmpOnlineCredentialsEmail
-                + "\","
-                + "\"code\":\""
-                + dmpOnlineCredentialsAuthorizationCode
-                + "\"}",
-            headers);
-    return restTemplate
-        .exchange(
-            new URL(dmpOnlineApiUrl + "/authenticate").toURI(),
-            HttpMethod.POST,
-            entity,
-            JsonNode.class)
-        .getBody()
-        .get("access_token")
-        .textValue();
+    if (StringUtils.isBlank(optUserConnection.get().getAccessToken())) {
+      throw new HttpClientErrorException(HttpStatus.NOT_FOUND, "Refresh token not found");
+    }
+    String accessToken;
+    if (!isConnectionAlive(principal)) {
+      refreshToken(model, principal);
+    }
+    accessToken = optUserConnection.get().getAccessToken();
+    return accessToken;
+  }
+
+  private String getExistingRefreshToken(Principal principal) {
+    Optional<UserConnection> optUserConnection = getExistingUserConnection(principal.getName());
+    if (optUserConnection.isEmpty()) {
+      throw new HttpClientErrorException(HttpStatus.NOT_FOUND, "User connection not found");
+    }
+    if (StringUtils.isBlank(optUserConnection.get().getRefreshToken())) {
+      throw new HttpClientErrorException(HttpStatus.NOT_FOUND, "Refresh token not found");
+    }
+    return optUserConnection.get().getRefreshToken();
   }
 
-  private HttpHeaders getHttpHeaders(String accessToken) {
+  private HttpHeaders getHttpHeadersWithToken(String accessToken) {
     HttpHeaders headers = new HttpHeaders();
     headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
-    headers.add("Authorization", String.format("Bearer %s", accessToken));
+    headers.setBearerAuth(accessToken);
     return headers;
   }
 
-  private static AjaxReturnObject<JsonNode> getErrorResponseOnNullToken() {
-    return new AjaxReturnObject<>(
-        null,
-        ErrorList.of(
-            "Couldn't generate DMPonline token. Is RSpace connection with DMPonline configured"
-                + " properly?"));
+  private URL getCallbackUrl() throws MalformedURLException {
+    if (StringUtils.isEmpty(callbackBaseUrl)) {
```

Comment:
yes, correct, `callbackBaseUrl` is optional and if the `callbackBaseUrl` is empty it will take `ServerUrl` as a base url for the callback.
I will specify it here https://ops2.researchspace.com/workspace/editor/structuredDocument/1820?

Reply:
I was wondering though if they will ever be different though? Won't it always be `{rspace-server-url}/apps/dmponline/callback` where rspace-server-url is `properties.getServerUrl`, or is it sometimes a different base url?

Reply:
I'll give you one of the example when they are not necessary be the same:

when you start it in local (for example), the host is on `localhost` but the callback needs to have the folloowing as prefix: `https://researchspace.eu.ngrok.io` 
We do not necessary need to change the server URL to that (that will be localhost in that case), we just need to change the callback url without changing the whole `rspace-server-url`

We have adopted this strategy for all the other integrations (DMPTool, DigitalCommonsData, and so on) that uses `OAuth2` protocols